### PR TITLE
[SPARK-27811][Core][Docs]Improve docs about spark.driver.memoryOverhead and spark.executor.memoryOverhead.

### PR DESCRIPTION
**Pull request by beliefer**
## What changes were proposed in this pull request?

I found the docs of `spark.driver.memoryOverhead` and `spark.executor.memoryOverhead` exists a little ambiguity.
For example, the origin docs of `spark.driver.memoryOverhead` start with `The amount of off-heap memory to be allocated per driver in cluster mode`.
But `MemoryManager` also managed a memory area named off-heap used to allocate memory in tungsten mode.
So I think the description of `spark.driver.memoryOverhead` always make confused.

`spark.executor.memoryOverhead` has the same confused with `spark.driver.memoryOverhead`.

## How was this patch tested?

Exists UT.

### CODE REVIEW

**Review by JoshRosen — CHANGES_REQUESTED** [review]
I think the proposed new description is inaccurate because interned strings and native overheads (e.g. Netty direct buffers) aren't allocated _outside the executor process_, per se (since off-heap memory is still in the JVM's own process address space).

Rather, I think the distinction is that these non-heap allocations don't count towards the JVM's total heap size limit, a factor which needs to be taken into account for container memory sizing: if you request a Spark executor with, say, a 4 gigabyte heap, then the actual peak memory usage of the process (from the OS's perspective) is going to be more than 4 gigabytes due to these types of off-heap allocations. 

If we want to avoid container OOM kills (by YARN or Kubernetes) then we need to account for this extra overhead somewhere, hence these `*memoryOverhead` "fudge-factors": setting the memory overhead causes us to request a container whose total memory size is greater than the heap size.

That said, increasing the memory overhead _does_ also result in additional memory headroom that can be used by non-driver/executor processes (like overhead from other processes in the container). Maybe we could state this explicitly, e.g. something like "... non-heap memory, including off-heap memory (e.g ....) and memory used by other non-driver / executor processes running in the same container".

You're correct that the relationship between these configurations and the Tungsten off-heap configuration is a bit confusing. To address that confusion, I'd prefer to expand the documentation to explicitly mention these `*memoryOverhead` configurations in the documentation for the Tungsten off-heap setting: I think that doc should recommend raising the `memoryOverhead` when setting the Tungsten config.

It might also help to more explicitly clarify that these settings only make sense in a containerized / resource limited deployment mode (e.g. not in standalone mode).

In summary, I think there's definitely room for confusion with the existing fix, but I think the right solution is an expansion of the docs to much more explicitly clarify the relationship between both sets of configurations, not a minor re-word.

**Comment by beliefer** [comment]
> I think the proposed new description is inaccurate because interned strings and native overheads (e.g. Netty direct buffers) aren't allocated _outside the executor process_, per se (since off-heap memory is still in the JVM's own process address space).

Thanks for your review. Yes, I ignored the detail off-heap memory is still in the JVM's own process address space.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by beliefer** [comment]
Retest this please.

**Comment by beliefer** [comment]
@JoshRosen I have changed.

**Review by beliefer — COMMENTED** [review]
review again!

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 1]
```diff
@@ -185,8 +189,12 @@ package object config {
     .createWithDefaultString("1g")
 
   private[spark] val EXECUTOR_MEMORY_OVERHEAD = ConfigBuilder("spark.executor.memoryOverhead")
-    .doc("The amount of off-heap memory to be allocated per executor in cluster mode, " +
-      "in MiB unless otherwise specified.")
+    .doc("Amount of non-heap memory to be allocated per executor process in cluster mode " +
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala]: I think that generally we reserve extended comments for the docs below. You don't need to change this.

File: docs/configuration.md [hunk 2]
```diff
@@ -181,10 +181,15 @@ of the most common options to set are:
...
   <td>
-    The amount of off-heap memory to be allocated per driver in cluster mode, in MiB unless
-    otherwise specified. This is memory that accounts for things like VM overheads, interned strings, 
-    other native overheads, etc. This tends to grow with the container size (typically 6-10%). 
-    This option is currently supported on YARN, Mesos and Kubernetes.
+    Amount of non-heap memory to be allocated per driver process in cluster mode 
+    (e.g YARN, Mesos and Kubernetes.), in MiB unless otherwise specified. This is memory that
+    accounts for things like VM overheads, interned strings, other native overheads, etc. 
+    This tends to grow with the container size (typically 6-10%). 
+    <em>Note:</em> Non-heap memory including off-heap memory
```
Comment (srowen) [inline docs/configuration.md]: Hm, what else gets counted towards the driver's memory here, in a YARN context? are you thinking of things like the Python process that goes with a Pyspark driver? I don't think this is wrong but at best that's the additional comment to make here.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala (hunk 1 above)
Comment (beliefer) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala]: OK. I revert comment in core, but reserve a little change.

File: docs/configuration.md (hunk 2 above)
Comment (beliefer) [reply docs/configuration.md]: Thanks for your reminder. I have extended comment about PySpark.

**Comment by beliefer** [comment]
Retest this please.

File: docs/configuration.md [hunk 3]
```diff
@@ -181,10 +181,16 @@ of the most common options to set are:
...
```
Comment (srowen) [inline docs/configuration.md]: including -> includes

File: docs/configuration.md [hunk 4]
```diff
@@ -1233,6 +1246,9 @@ Apart from these, the following properties are also available, and may be useful
   <td>
     If true, Spark will attempt to use off-heap memory for certain operations. If off-heap memory 
     use is enabled, then <code>spark.memory.offHeap.size</code> must be positive.
+    <em>Note:</em> If off-heap memory use is enabled or off-heap memory size is increased, 
+    recommend raising the non-heap memory size(e.g increase <code>spark.driver.memoryOverhead</code>
```
Comment (srowen) [inline docs/configuration.md]: recommend raising -> raise

File: docs/configuration.md [hunk 5]
```diff
@@ -1233,6 +1246,9 @@ Apart from these, the following properties are also available, and may be useful
   <td>
     If true, Spark will attempt to use off-heap memory for certain operations. If off-heap memory 
     use is enabled, then <code>spark.memory.offHeap.size</code> must be positive.
+    <em>Note:</em> If off-heap memory use is enabled or off-heap memory size is increased,
```
Comment (srowen) [inline docs/configuration.md]: What is 'off-heap memory size is increased' referring to?

**Review by srowen — CHANGES_REQUESTED** [review]

File: docs/configuration.md (hunk 3 above)
Comment (beliefer) [reply docs/configuration.md]: OK. I changed.

File: docs/configuration.md (hunk 5 above)
Comment (beliefer) [reply docs/configuration.md]: increase the size of off-heap memory by `spark.memory.offHeap.size`.

File: docs/configuration.md (hunk 4 above)
Comment (beliefer) [reply docs/configuration.md]: Thanks. I changed.

File: docs/configuration.md (hunk 5 above)
Comment (srowen) [reply docs/configuration.md]: This is implied by enabling off-heap memory; I'd just remove the second part of this statement.
Below: remove the word 'recommend'. Space before the paren and fix "e.g."
But isn't this already accounted for by `spark.memory.offHeap.size`?

Comment (beliefer) [reply docs/configuration.md]: OK, I have adjusted the documentation according to your reminder.
When Spark running in cluster mode (e.g. YARN), the size `spark.memory.offHeap.size` is not accounted in total memory size of container. As size of off-heap memory grows, increase the size of the non-heap memory manually.

File: docs/configuration.md [hunk 6]
```diff
@@ -181,10 +181,16 @@ of the most common options to set are:
   <td><code>spark.driver.memoryOverhead</code></td>
   <td>driverMemory * 0.10, with minimum of 384 </td>
   <td>
-    The amount of off-heap memory to be allocated per driver in cluster mode, in MiB unless
-    otherwise specified. This is memory that accounts for things like VM overheads, interned strings, 
-    other native overheads, etc. This tends to grow with the container size (typically 6-10%). 
-    This option is currently supported on YARN, Mesos and Kubernetes.
+    Amount of non-heap memory to be allocated per driver process in cluster mode 
+    (e.g. YARN, Mesos and Kubernetes.), in MiB unless otherwise specified. This is memory that
```
Comment (srowen) [inline docs/configuration.md]: This line changed, to mention YARN et al here. It implies that these resource managers always run in cluster mode. They don't, so I would revert this change and restore the sentence starting with "This option..."

File: docs/configuration.md [hunk 7]
```diff
@@ -1233,6 +1246,8 @@ Apart from these, the following properties are also available, and may be useful
   <td>
     If true, Spark will attempt to use off-heap memory for certain operations. If off-heap memory 
     use is enabled, then <code>spark.memory.offHeap.size</code> must be positive.
+    <em>Note:</em> If off-heap memory is enabled, raise the non-heap memory size (e.g. increase
```
Comment (srowen) [inline docs/configuration.md]: OK, rather than say "raise", I think the idea here is to suggest that you _may need to raise_ these values. You may not need to.

**Review by srowen — CHANGES_REQUESTED** [review]

File: docs/configuration.md (hunk 6 above)
Comment (beliefer) [reply docs/configuration.md]: OK , I have reverted "This option..." and delete "(e.g. "

File: docs/configuration.md (hunk 7 above)
Comment (beliefer) [reply docs/configuration.md]: I added "you may need to" as a suggestion.

**Review by srowen — APPROVED** [review]

**Comment by beliefer** [comment]
@JoshRosen Could you review this PR again and find other issues?

**Comment by srowen** [comment]
Merged to master

**Comment by beliefer** [comment]
@srowen Thanks for your merger. I thought that two change requested need every one to agree.
@JoshRosen Thanks for your detailed review.
